File: ios/RNEncryptedStorage.m
//
//  RNEncryptedStorage.m
//  Starter
//
//  Created by Yanick Bélanger on 2020-02-09.
//  Copyright © 2020 Facebook. All rights reserved.
//

#import "RNEncryptedStorage.h"
#import <Security/Security.h>
#import <React/RCTLog.h>

void rejectPromise(NSString *message, NSError *error, RCTPromiseRejectBlock rejecter)
{
    NSString* errorCode = [NSString stringWithFormat:@"%ld", error.code];
    NSString* errorMessage = [NSString stringWithFormat:@"RNEncryptedStorageError: %@", message];

    rejecter(errorCode, errorMessage, error);
}

@implementation RNEncryptedStorage

+ (BOOL)requiresMainQueueSetup
{
    return NO;
}

CFStringRef getKeychainAccessibility(NSDictionary *options)
{
    id keychainAccessibility = options[@"keychainAccessibility"];
    if (keychainAccessibility == nil) {
        return kSecAttrAccessibleAfterFirstUnlock;
    }
        
    NSDictionary *valueMap = @{
        @"kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly": (__bridge NSString *)kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly,
        @"kSecAttrAccessibleWhenUnlockedThisDeviceOnly": (__bridge NSString *)kSecAttrAccessibleWhenUnlockedThisDeviceOnly,
        @"kSecAttrAccessibleWhenUnlocked": (__bridge NSString *)kSecAttrAccessibleWhenUnlocked,
        @"kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly": (__bridge NSString *)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        @"kSecAttrAccessibleAfterFirstUnlock": (__bridge NSString *)kSecAttrAccessibleAfterFirstUnlock,
        
    };
    
    NSString *value = valueMap[keychainAccessibility];
    
    return (__bridge CFStringRef)value;
}

NSString *getKeychainService(NSDictionary *options)
{
    id keychainService = options[@"storageName"];
    if (keychainService == nil) {
        return [[NSBundle mainBundle] bundleIdentifier];
    }
    return keychainService;
}


RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(setItem:(NSString *)key withValue:(NSString *)value withOptions:(NSDictionary *) options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    NSData* dataFromValue = [value dataUsingEncoding:NSUTF8StringEncoding];
    
    if (dataFromValue == nil) {
        NSError* error = [NSError errorWithDomain:[[NSBundle mainBundle] bundleIdentifier] code:0 userInfo: nil];
        rejectPromise(@"An error occured while parsing value", error, reject);
        return;
    }
    
    // Prepares the insert query structure
    CFStringRef keychainAccessibility = getKeychainAccessibility(options);
    NSString *keychainService = getKeychainService(options);
    NSDictionary* storeQuery = @{
        (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrAccount : key,
        (__bridge id)kSecValueData : dataFromValue,
        (__bridge id)kSecAttrAccessible: (__bridge id)keychainAccessibility,
        (__bridge id)kSecAttrService: keychainService
    };
    
    // Deletes the existing item prior to inserting the new one
    SecItemDelete((__bridge CFDictionaryRef)storeQuery);
    
    OSStatus insertStatus = SecItemAdd((__bridge CFDictionaryRef)storeQuery, nil);
    
    if (insertStatus == noErr) {
        resolve(value);
    }
    
    else {
        NSError* error = [NSError errorWithDomain:[[NSBundle mainBundle] bundleIdentifier] code:insertStatus userInfo: nil];
        rejectPromise(@"An error occured while saving value", error, reject);   
    }
}

RCT_EXPORT_METHOD(getItem:(NSString *)key withOptions:(NSDictionary *) options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    CFStringRef keychainAccessibility = getKeychainAccessibility(options);
    NSString *keychainService = getKeychainService(options);
    NSDictionary* getQuery = @{
        (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrAccount : key,
        (__bridge id)kSecReturnData : (__bridge id)kCFBooleanTrue,
        (__bridge id)kSecMatchLimit : (__bridge id)kSecMatchLimitOne,
        (__bridge id)kSecAttrAccessible: (__bridge id)keychainAccessibility,
        (__bridge id)kSecAttrService: keychainService
    };
    
    CFTypeRef dataRef = NULL;
    OSStatus getStatus = SecItemCopyMatching((__bridge CFDictionaryRef)getQuery, &dataRef);
    
    if (getStatus == errSecSuccess) {
        NSString* storedValue = [[NSString alloc] initWithData: (__bridge NSData*)dataRef encoding: NSUTF8StringEncoding];
        resolve(storedValue);
    }

    else if (getStatus == errSecItemNotFound) {
        resolve(nil);
    }
    
    else {
        NSError* error = [NSError errorWithDomain: [[NSBundle mainBundle] bundleIdentifier] code:getStatus userInfo:nil];
        rejectPromise(@"An error occured while retrieving value", error, reject);
    }
}

RCT_EXPORT_METHOD(removeItem:(NSString *)key withOptions:(NSDictionary *) options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    CFStringRef keychainAccessibility = getKeychainAccessibility(options);
    NSString *keychainService = getKeychainService(options);
    NSDictionary* removeQuery = @{
        (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecAttrAccount : key,
        (__bridge id)kSecReturnData : (__bridge id)kCFBooleanTrue,
        (__bridge id)kSecAttrAccessible: (__bridge id)keychainAccessibility,
        (__bridge id)kSecAttrService: keychainService
    };
    
    OSStatus removeStatus = SecItemDelete((__bridge CFDictionaryRef)removeQuery);
    
    if (removeStatus == noErr) {
        resolve(key);
    }
    
    else {
        NSError* error = [NSError errorWithDomain:[[NSBundle mainBundle] bundleIdentifier] code:removeStatus userInfo: nil];
        rejectPromise(@"An error occured while removing value", error, reject);
    }
}

RCT_EXPORT_METHOD(clear:(NSDictionary *) options resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)
{
    NSArray *secItemClasses = @[
        (__bridge id)kSecClassGenericPassword,
        (__bridge id)kSecClassInternetPassword,
        (__bridge id)kSecClassCertificate,
        (__bridge id)kSecClassKey,
        (__bridge id)kSecClassIdentity
    ];
    
    NSString *keychainService = getKeychainService(options);
    
    // Maps through all Keychain classes and deletes all items that match
    for (id secItemClass in secItemClasses) {
        NSDictionary *spec = @{
            (__bridge id)kSecClass: secItemClass,
            (__bridge id)kSecAttrService: keychainService
        };
        SecItemDelete((__bridge CFDictionaryRef)spec);
    }
    
    resolve(nil);
}
@end

File: ios/RNEncryptedStorage.swift
//
//  RNEncryptedStorage.swift
//  EncryptedStorage
//
//  Created by Yanick Bélanger on 2020-02-11.
//  Copyright © 2020 Facebook. All rights reserved.
//

import Foundation
import Security

@objc(RNEncryptedStorage)
class RNEncryptedStorage: NSObject {
  
  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false;
  }
  
  @objc func store(_ key : String, value : String, resolver resolve : RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) -> Void {
    guard let dataFromValue = value.data(using: .utf8, allowLossyConversion: false) else {
      return reject("Error parsing value for key \(key)", nil, nil);
    }
    
    // Prepares the insert query structure
    let storeQuery : [String : Any] = [
      kSecClass as String : kSecClassGenericPassword,
      kSecAttrAccount as String : key,
      kSecValueData as String : dataFromValue
    ];
    
    // Deletes the existing item prior to inserting the new one
    SecItemDelete(storeQuery as CFDictionary);
    
    let status = SecItemAdd(storeQuery as CFDictionary, nil);
    
    if (status == noErr) {
      resolve(value);
    }
    
    else {
      reject("An error occured while saving \(key)", nil, nil);
    }
  }
  
  @objc
  func retrieve(_ key : String, resolver resolve : RCTPromiseResolveBlock, rejecter reject : RCTPromiseRejectBlock) {
    let retrieveQuery : [String : Any] = [
      kSecClass as String : kSecClassGenericPassword,
      kSecAttrAccount as String : key,
      kSecReturnData as String : kCFBooleanTrue!,
      kSecMatchLimit as String : kSecMatchLimitOne
    ];
    
    var dataRef : AnyObject? = nil;
    
    let status = SecItemCopyMatching(retrieveQuery as CFDictionary, &dataRef);
    
    if (status == noErr) {
      let stringValue = String(data : (dataRef as! Data?)!, encoding: .utf8)!;
      resolve(stringValue);
    }
    
    else {
      reject("An error occured while retrieving \(key)", nil, nil);
    }
  }
  
  @objc
  func remove(_ key : String, resolver resolve : RCTPromiseResolveBlock, rejecter reject : RCTPromiseRejectBlock) {
    let removeQuery : [String : Any] = [
      kSecClass as String : kSecClassGenericPassword,
      kSecAttrAccount as String : key,
      kSecReturnData as String : kCFBooleanTrue!
    ];
    
    let status = SecItemDelete(removeQuery as CFDictionary);
    
    if (status == noErr) {
      resolve(key);
    }
    
    else {
      reject("An error occured while removing \(key)", nil, nil);
    }
  }
}

